### PR TITLE
Add support for delegated KT accounts with active default entrypoint

### PR DESCRIPTION
**Pull request by amzid**
This PR implements support for delegated KT accounts with active default entrypoints. The TRD calculates the simulated consumed gas and sets the limit for gas and the baker fee on the basis of the calculated amount.

This PR is a partial fix for #352: It covers the delegated contracts which have a default entrypoint. If the delegated account does not have a default entrypoint, the user should forward the payment to another supported payment address using the rules_map.

UPDATE: add storage cost computation + bound total fee amount.

* **Performed tests**: Test payouts to vaults on Delphinet.

**Work effort**: 7 hrs

### CODE REVIEW

**Comment by amzid** [comment]
A review from the side of @maht0rz suggests the following:
1) the implementation should cover potential extra storage requirements
2) We should assert that the resulting fee (with extra gas and storage costs) should remain below the reward amount.
3) Reward amounts for multiple cycles can be aggregated to have a greater value than the required fee. This is even more important for KT1 accounts.
4) With the current gas estimation for vaults and the current block gas hard limit for delphi, the batch size cannot exceed 100. The batch size would be should be computed dynamically,

The first two suggesions will be implemented in this PR. For the two remaining points (3 & 4), extra issues will be opened.

**Review by utdrmac — APPROVED** [review]

**Comment by dexterslabor** [comment]
Hi @amzid, at Stove Labs we are developing an SDK for interacting with vaults/ovens of the wXTZ project. 
It has a method implemented to check for contract code integrity. Maybe this is of any use to identify correctly oven contracts.

Fetching
https://github.com/stove-labs/wrapped-xtz-sdk/blob/dev/src/WXTZBase.ts#L34-L39
Checksum
https://github.com/stove-labs/wrapped-xtz-sdk/blob/dev/src/deployments.json#L24
Checking
https://github.com/stove-labs/wrapped-xtz-sdk/blob/dev/src/byteHelpers.ts#L17-L20

**Comment by jdsika** [comment]
@amzid is this ready to be reviewed now? Are all comments considered?

**Comment by amzid** [comment]
#357 is opened for remaining work

File: src/pay/batch_payer.py [hunk 1]
```diff
@@ -292,13 +347,26 @@ def attempt_single_batch(self, payment_records, op_counter, dry_run=None):
...
 
+            # TRD extension for non scriptless contract accounts
+            gas_limit, tx_fee = self.gas_limit, self.default_fee
+            if payment_item.paymentaddress.startswith('KT'):
+                simulation_results = self.simulate_single_operation(payment_item, pymnt_amnt, branch, chain_id)
+                gas_limit, tx_fee, storage = simulation_results
+                burn_fee = COST_PER_BYTE * storage
+                total_fee = tx_fee + burn_fee
+                # Bound the total (baker and burn) fee by the reward amount in case of KT1 accounts
+                if total_fee > pymnt_amnt:
```
Comment (jdsika) [inline src/pay/batch_payer.py:358]: where is the connection to the limits set in the fee.ini?

Comment (amzid) [reply src/pay/batch_payer.py:358]: The initial fee value is taken from fee.ini. The extra fee needed for gas + extra burn fee needed for storage are simulated and adjusted according to the simulation output. The values for tz1 accounts are taken directly from fee.ini.

Comment (jdsika) [reply src/pay/batch_payer.py:358]: the boundary if(fee > reward) is not sufficient. we need maximum limits services are willing to pay. Idea: create the fee.ini at first startup. Add the name of the current protocol. check it at startup and update with new fees if protocol changes. Store the simulated values for a tz account and add max_fees to the ini as a fail criteria. The Kt accounts with internal calls can be used to trick services into paying large amounts of fees.

**Comment by jdsika** [comment]
Initial state: fee.ini is not part of the repository
1) Check if fee.ini exists
-> no: create fee.ini, fetch hard limits from protocol, fetch protocol verison, simulate tz account costs and write to fee.ini
-> yes: check if fee.ini protocol version is the same as the currently used one. -> if no: delete fee.ini and initialize again

**Comment by jdsika** [comment]
@amzid I think you still need to change the parts we discused before @utdrmac can do a final approval?

**Comment by amzid** [comment]
> @amzid I think you still need to change the parts we discused before @utdrmac can do a final approval?

I would like to close this PR here. This includes the capability of paying out to KT1 contracts with default entrypoints with a fee limit adjusted on basis of the oven code (which should be the most costly KT contract known to TRD up to now). The identification of KT accounts and adjustable fees according to protocol/contract is opened in a separate issue described in #357.

**Review by jdsika — APPROVED** [review]
LGTM

**Pull request merged**